core/src/main/java/org/apache/calcite/rex/RexSimplify.java

```diff
@@ -1624,6 +1632,49 @@ private RexNode simplifyOrs(List<RexNode> terms, RexUnknownAs unknownAs) {
             continue;
           }
         }
+        break;
+      case NOT_EQUALS:
+        final Comparison notEqualsComparison =
+            Comparison.of(term, node -> RexUtil.isReferenceOrAccess(node, false));
+        if (notEqualsComparison != null) {
+          // We are dealing with a X<>A term, check if we saw before another NOT_EQUALS involving X
+          final RexNode prevNotEquals = notEqualsComparisonMap.get(notEqualsComparison.ref);
+          if (prevNotEquals == null) {
+            // This is the first NOT_EQUALS involving X, put it in the map
+            notEqualsComparisonMap.put(notEqualsComparison.ref, term);
+          } else {
+            // There is already in the map another NOT_EQUALS involving X:
+            //   - if it is already an IS_NOT_NULL: it was already simplified, ignore this term
+            //   - if it is not an IS_NOT_NULL (i.e. it is a NOT_EQUALS): check comparison values
+            if (prevNotEquals.getKind() != SqlKind.IS_NOT_NULL) {
+              final Comparable comparable1 = notEqualsComparison.literal.getValue();
+              //noinspection ConstantConditions
+              final Comparable comparable2 = Comparison.of(
+                  prevNotEquals, node -> RexUtil.isReferenceOrAccess(node, false))
```

Comment:
any particular reason why this is not just `Comparison.of(prevNotEquals)` ?
we can exclude casts; but I don't see any problem keeping them in this case

Reply:
Uhm, that was my initial idea but I think I had to exclude casts because it caused some problems.... but now I'm testing with them and tests seem fine, so it seems I was wrong. Anyway, I will commit the change to use `Comparison of(RexNode e)`

Reply:
Change committed to consider casts